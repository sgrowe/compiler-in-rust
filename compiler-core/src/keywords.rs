use std::fmt;

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum Keyword {
    Import,
    Export,
    Type,
    Function,
}

pub fn get_matching_keyword(name: &str) -> Option<Keyword> {
    use Keyword::*;

    let keyword = match name {
        "import" => Import,
        "export" => Export,
        "type" => Type,
        "fn" => Function,
        _ => return None,
    };

    Some(keyword)
}

impl Keyword {
    pub fn as_str(self) -> &'static str {
        use Keyword::*;

        match self {
            Import => "import",
            Export => "export",
            Type => "type",
            Function => "fn",
        }
    }
}

impl fmt::Display for Keyword {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.as_str())
    }
}
