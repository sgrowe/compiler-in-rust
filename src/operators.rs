pub trait BindingPower {
    fn binding_power(&self) -> u32;
}

#[derive(Debug, Copy, Clone)]
pub enum BinaryOperator {
    Plus,
    Minus,
    Multiply,
}

impl BindingPower for BinaryOperator {
    fn binding_power(&self) -> u32 {
        use BinaryOperator::*;

        match self {
            Plus => 50,
            Minus => 50,
            Multiply => 60,
        }
    }
}
