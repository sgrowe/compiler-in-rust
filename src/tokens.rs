#[derive(Debug, Copy, Clone)]
pub enum Token<'a> {
    RightArrow,
    Equals,
    Pipe,
    Name(&'a str),
    Keyword(Keyword),
    BinOp(BinaryOperator),
    Constant(Constant<'a>),
}

#[derive(Debug, Copy, Clone)]
pub enum BinaryOperator {
    Plus,
    Minus,
}

#[derive(Debug, Copy, Clone)]
pub enum Constant<'a> {
    Str(&'a str),
    Float(f64),
    Int(i64),
}

#[derive(Debug, Copy, Clone)]
pub enum Keyword {
    Import,
    Export,
    Type,
    Function,
}
