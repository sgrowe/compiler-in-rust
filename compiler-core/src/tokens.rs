use super::keywords::Keyword;
use super::operators::BinaryOperator;

#[derive(Debug, Copy, Clone)]
pub enum Token<'a> {
    Equals,
    Pipe,
    Comma,
    Colon,
    FatRightArrow,
    OpenParen,
    CloseParen,
    IndentIncr,
    IndentDecr,
    Name(&'a str),
    Keyword(Keyword),
    BinOp(BinaryOperator),
    Constant(Constant<'a>),
}

#[derive(Debug, Copy, Clone)]
pub enum Constant<'a> {
    Str(&'a str),
    Float(f64),
    Int(i64),
}
