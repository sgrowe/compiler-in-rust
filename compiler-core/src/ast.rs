use super::operators::*;
use super::tokens::*;

#[derive(Debug, Clone)]
pub struct Ast<'a> {
    pub statements: Vec<TopLevelStatement<'a>>,
}

impl<'a> Ast<'a> {
    pub fn new() -> Ast<'a> {
        Ast {
            statements: Vec::new(),
        }
    }

    pub fn append_statement(&mut self, statement: TopLevelStatement<'a>) {
        self.statements.push(statement)
    }
}

#[derive(Debug, Clone)]
pub enum Declaration<'a> {
    Assignment {
        name: &'a str,
        expr: Expression<'a>,
    },
    FunctionDecl {
        name: &'a str,
        arguments: FunctionArgsList<'a>,
        body: Vec<FuncBodyStatement<'a>>,
    },
}

#[derive(Debug, Clone)]
pub enum TopLevelStatement<'a> {
    Declaration {
        decl: Declaration<'a>,
        exported: bool,
    },
}

#[derive(Debug, Clone)]
pub enum FuncBodyStatement<'a> {
    Declaration(Declaration<'a>),
    BareExpression(Expression<'a>),
}

#[derive(Debug, Clone)]
pub struct FunctionArgsList<'a> {
    pub args: Vec<FunctionArg<'a>>,
}

#[derive(Debug, Clone)]
pub struct FunctionArg<'a> {
    pub name: &'a str,
}

#[derive(Debug, Clone)]
pub enum Expression<'a> {
    Variable(&'a str),
    Constant(Constant<'a>),
    FunctionCall {
        name: &'a str,
        args: Vec<Expression<'a>>,
    },
    BinaryOp {
        operator: BinaryOperator,
        left: Box<Expression<'a>>,
        right: Box<Expression<'a>>,
    },
    Negation(Box<Expression<'a>>),
}
