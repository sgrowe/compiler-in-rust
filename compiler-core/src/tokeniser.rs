use super::keywords::*;
use super::tokens::*;
use std::iter::Peekable;
use std::str::CharIndices;

use std::cmp::Ordering;

pub type Result<X> = std::result::Result<X, TokeniserError>;

pub fn tokenise<'a>(source: &'a str) -> Tokeniser<'a> {
    Tokeniser {
        source,
        chars: source.char_indices().peekable(),
        indent_stack: Vec::with_capacity(8),
    }
}

#[derive(Debug, Clone)]
pub struct Tokeniser<'a> {
    source: &'a str,
    chars: Peekable<CharIndices<'a>>,
    indent_stack: Vec<u16>, // u16 as hopefully people wont use many more that 65,000 spaces of indentation
}

impl<'a> Iterator for Tokeniser<'a> {
    type Item = Result<Token<'a>>;

    fn next(&mut self) -> Option<Self::Item> {
        use super::operators::BinaryOperator::*;
        use Token::*;

        while let Some((i, c)) = self.step() {
            match c {
                ' ' => {
                    continue;
                }
                '\n' | '\r' => {
                    if let Some(token) = self.newline() {
                        return Some(Ok(token));
                    }

                    continue;
                }
                _ => {}
            }

            let token = match c {
                '(' => OpenParen,
                ')' => CloseParen,
                '"' => match self.string_constant(i) {
                    Ok(str_const) => str_const,
                    Err(error) => return Some(Err(error)),
                },
                '|' => Pipe,
                '+' => BinOp(Plus),
                '*' => BinOp(Multiply),
                '-' => BinOp(Minus),
                '/' => BinOp(Divide),
                ',' => Comma,
                '=' => match self.peek_next_char() {
                    Some('>') => {
                        self.step();
                        FatRightArrow
                    }
                    _ => Equals,
                },
                _ => {
                    if c.is_alphabetic() {
                        self.name(i)
                    } else if c.is_numeric() {
                        self.number(i)
                    } else {
                        return None;
                    }
                }
            };

            return Some(Ok(token));
        }

        if self.indent_stack.pop().is_some() {
            Some(Ok(IndentDecr))
        } else {
            None
        }
    }
}

impl<'a> Tokeniser<'a> {
    fn step(&mut self) -> Option<(usize, char)> {
        self.chars.next()
    }

    fn peek_next_char(&mut self) -> Option<char> {
        self.chars.peek().map(|&(_, c)| c)
    }

    fn current_indent(&self) -> u16 {
        self.indent_stack.last().copied().unwrap_or(0)
    }

    fn newline(&mut self) -> Option<Token<'a>> {
        use Token::{IndentDecr, IndentIncr};

        let mut indent = 0;

        while let Some(c) = self.peek_next_char() {
            match c {
                ' ' => {
                    indent += 1;
                }
                '\n' | '\r' => {
                    // ignore blank lines
                    indent = 0;
                }
                _ => {
                    return match indent.cmp(&self.current_indent()) {
                        Ordering::Greater => {
                            self.indent_stack.push(indent);
                            Some(IndentIncr)
                        }
                        Ordering::Less => {
                            self.indent_stack.pop();
                            Some(IndentDecr)
                        }
                        Ordering::Equal => None,
                    };
                }
            }

            self.step();
        }

        None
    }

    fn string_constant(&mut self, start: usize) -> Result<Token<'a>> {
        while let Some((i, c)) = self.step() {
            if c == '"' {
                return Ok(Token::Constant(Constant::Str(&self.source[start + 1..i])));
            }
        }

        Err(TokeniserError::UnterminatedString)
    }

    fn name(&mut self, start: usize) -> Token<'a> {
        let mut end = start + 1;

        while let Some(&(i, c)) = self.chars.peek() {
            end = i;

            if c.is_alphabetic() || c == '_' {
                self.step();
            } else {
                break;
            }
        }

        let name = &self.source[start..end];

        get_matching_keyword(name)
            .map(Token::Keyword)
            .unwrap_or(Token::Name(name))
    }

    fn number(&mut self, start: usize) -> Token<'a> {
        let mut end = start + 1;
        let mut is_float = false;

        while let Some(&(i, c)) = self.chars.peek() {
            end = i;

            if c.is_numeric() {
                self.step();
            } else if !is_float && c == '.' {
                is_float = true;
                self.step();
            } else {
                break;
            }
        }

        let num = &self.source[start..end];

        let constant = if is_float {
            Constant::Float(num.parse().unwrap())
        } else {
            Constant::Int(num.parse().unwrap())
        };

        Token::Constant(constant)
    }
}

#[derive(Debug, Copy, Clone)]
pub enum TokeniserError {
    UnterminatedString,
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use std::fs;
    use test_case::test_case;

    #[test_case("src/fixtures/strings.lang"; "strings")]
    #[test_case("src/fixtures/maths.lang"; "maths")]
    #[test_case("src/fixtures/functions.lang"; "functions")]
    fn fixtures(fixture_file_name: &str) -> std::io::Result<()> {
        let contents = fs::read_to_string(fixture_file_name)?;

        let tokens = tokenise(&contents).collect::<Vec<_>>();

        assert_debug_snapshot!(tokens);

        Ok(())
    }
}
