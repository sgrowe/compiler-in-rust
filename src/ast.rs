use super::operators::*;
use super::tokens::*;

#[derive(Debug, Clone)]
pub struct Ast<'a> {
    pub statements: Vec<Statement<'a>>,
}

impl<'a> Ast<'a> {
    pub fn append_statement(&mut self, statement: Statement<'a>) {
        self.statements.push(statement)
    }
}

#[derive(Debug, Clone)]
pub enum Statement<'a> {
    Assignment {
        name: &'a str,
        expr: Expression<'a>,
    },
    BareExpression(Expression<'a>),
    FunctionDecl {
        name: &'a str,
        arguments: FunctionArgsList<'a>,
        body: Vec<Statement<'a>>,
    },
}

#[derive(Debug, Clone)]
pub struct FunctionArgsList<'a> {
    pub args: Vec<FunctionArg<'a>>,
}

#[derive(Debug, Clone)]
pub struct FunctionArg<'a> {
    pub name: &'a str,
}

#[derive(Debug, Clone)]
pub enum Expression<'a> {
    Variable(&'a str),
    Constant(Constant<'a>),
    BinaryOp {
        operator: BinaryOperator,
        left: Box<Expression<'a>>,
        right: Box<Expression<'a>>,
    },
}
