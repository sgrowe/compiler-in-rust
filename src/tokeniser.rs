use super::tokens::*;
use std::str::CharIndices;

pub fn tokenise<'a>(source: &'a str) -> Tokeniser<'a> {
    let mut chars = source.char_indices();
    let next_char = chars.next();

    Tokeniser {
        source,
        next_char,
        chars,
    }
}

#[derive(Debug, Clone)]
pub struct Tokeniser<'a> {
    source: &'a str,
    chars: CharIndices<'a>,
    // the next char in the sequence so we can lookahead one char without consuming the iterator
    next_char: Option<(usize, char)>,
}

impl<'a> Iterator for Tokeniser<'a> {
    type Item = Result<Token<'a>, TokeniserError>;

    fn next(&mut self) -> Option<Self::Item> {
        use self::BinaryOperator::*;
        use Token::*;
        use TokeniserError::*;

        while let Some((i, c)) = self.step() {
            if c.is_whitespace() {
                continue;
            }

            return match (c, self.peek_next_char()) {
                ('"', _) => Some(self.string_constant(i)),
                ('|', _) => Some(Ok(Pipe)),
                ('=', Some('>')) => {
                    self.step();
                    Some(Ok(RightArrow))
                }
                ('=', _) => Some(Ok(Equals)),
                ('+', _) => Some(Ok(BinOp(Plus))),
                ('-', _) => Some(Ok(BinOp(Minus))),
                (_, _) => {
                    if c.is_alphabetic() {
                        Some(Ok(self.name(i)))
                    } else if c.is_numeric() {
                        Some(Ok(self.number(i)))
                    } else {
                        None
                    }
                }
            };
        }

        None
    }
}

fn get_matching_keyword(name: &str) -> Option<Keyword> {
    use Keyword::*;

    match name {
        "import" => Some(Import),
        "export" => Some(Export),
        "type" => Some(Type),
        "fn" => Some(Function),
        _ => None,
    }
}

impl<'a> Tokeniser<'a> {
    fn step(&mut self) -> Option<(usize, char)> {
        let current = self.next_char;
        self.next_char = self.chars.next();
        current
    }

    fn peek_next_char(&self) -> Option<char> {
        self.next_char.map(|(_, c)| c)
    }

    fn string_constant(&mut self, start: usize) -> Result<Token<'a>, TokeniserError> {
        while let Some((i, c)) = self.step() {
            if c == '"' {
                return Ok(Token::Constant(Constant::Str(&self.source[start + 1..i])));
            }
        }

        Err(TokeniserError::UnterminatedString)
    }

    fn name(&mut self, start: usize) -> Token<'a> {
        let mut end = start + 1;

        while let Some((i, c)) = self.next_char {
            end = i;

            if c.is_alphabetic() || c == '_' {
                self.step();
            } else {
                break;
            }
        }

        let name = &self.source[start..end];

        get_matching_keyword(name)
            .map(|keyword| Token::Keyword(keyword))
            .unwrap_or(Token::Name(name))
    }

    fn number(&mut self, start: usize) -> Token<'a> {
        let mut end = start + 1;
        let mut is_float = false;

        while let Some((i, c)) = self.next_char {
            end = i;

            if c.is_numeric() {
                self.step();
            } else if !is_float && c == '.' {
                is_float = true;
                self.step();
            } else {
                break;
            }
        }

        let num = &self.source[start..end];

        let constant = if is_float {
            Constant::Float(num.parse().unwrap())
        } else {
            Constant::Int(num.parse().unwrap())
        };

        Token::Constant(constant)
    }
}

#[derive(Debug, Copy, Clone)]
pub enum TokeniserError {
    UnterminatedString,
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use std::fs;
    use test_case::test_case;

    #[test_case("src/fixtures/strings.lang"; "tokenise strings")]
    #[test_case("src/fixtures/maths.lang"; "tokenise maths expressions")]
    fn fixtures(fixture_file_name: &str) -> std::io::Result<()> {
        let contents = fs::read_to_string(fixture_file_name)?;

        let tokens = tokenise(&contents).collect::<Vec<_>>();

        assert_debug_snapshot!(tokens);

        Ok(())
    }
}
