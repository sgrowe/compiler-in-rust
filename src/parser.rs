use super::ast::*;
use super::operators::*;
use super::tokeniser::*;
use super::tokens::Token::*;
use super::tokens::*;

pub fn parse<'a>(source: &'a str) -> Result<Ast<'a>, ParseError<'a>> {
    Parser::of(tokenise(source)).parse()
}

#[derive(Debug, Clone)]
struct Parser<'a> {
    next_token: Result<Option<Token<'a>>, ParseError<'a>>,
    tokens: Tokeniser<'a>,
}

fn next_token<'a>(tokens: &mut Tokeniser<'a>) -> Result<Option<Token<'a>>, ParseError<'a>> {
    match tokens.next() {
        Some(result) => result
            .map_err(ParseError::TokeniserError)
            .map(|token| Some(token)),
        None => Ok(None),
    }
}

impl<'a> Parser<'a> {
    fn of(mut tokens: Tokeniser<'a>) -> Parser<'a> {
        Parser {
            next_token: next_token(&mut tokens),
            tokens,
        }
    }

    fn step(&mut self) -> Result<Option<Token<'a>>, ParseError<'a>> {
        let current = self.next_token;
        self.next_token = next_token(&mut self.tokens);
        current
    }

    fn peek_next_token(&self) -> Result<Option<Token<'a>>, ParseError<'a>> {
        self.next_token
    }

    fn parse(&mut self) -> Result<Ast<'a>, ParseError<'a>> {
        let mut ast = Ast {
            statements: Vec::new(),
        };

        while let Some(_) = self.peek_next_token()? {
            ast.append_statement(self.statement()?);
        }

        Ok(ast)
    }

    fn statement(&mut self) -> Result<Statement<'a>, ParseError<'a>> {
        use super::keywords::Keyword::*;

        if let Some(token) = self.step()? {
            match (token, self.peek_next_token()?) {
                (Name(name), _) => self.named_statement(name),
                (Keyword(Function), _) => self.function(),
                (Constant(_), _) => Ok(Statement::BareExpression(self.expression(0, Some(token))?)),
                (token, _) => Err(ParseError::UnexpectedToken(token)),
            }
        } else {
            Err(ParseError::UnexpectedEndOfInput)
        }
    }

    fn named_statement(&mut self, name: &'a str) -> Result<Statement<'a>, ParseError<'a>> {
        use Statement::*;

        match self.peek_next_token()? {
            Some(Equals) => {
                self.step()?;

                Ok(Assignment {
                    name,
                    expr: self.expression(0, None)?,
                })
            }
            Some(IndentDecr) => Ok(BareExpression(Expression::Variable(name))),
            Some(_) => Ok(BareExpression(self.expression(0, Some(Token::Name(name)))?)),
            None => Ok(BareExpression(Expression::Variable(name))),
        }
    }

    fn function(&mut self) -> Result<Statement<'a>, ParseError<'a>> {
        match self.step()? {
            Some(Name(name)) => {
                let arguments = self.function_arguments_list()?;
                let body = self.function_body()?;

                Ok(Statement::FunctionDecl {
                    name,
                    arguments,
                    body,
                })
            }
            _ => Err(ParseError::FunctionParseError),
        }
    }

    fn function_arguments_list(&mut self) -> Result<FunctionArgsList<'a>, ParseError<'a>> {
        match self.step()? {
            Some(OpenParen) => {
                let mut args = Vec::new();

                while let Some(token) = self.step()? {
                    match token {
                        Name(name) => args.push(self.func_arg(name)?),
                        CloseParen => return Ok(FunctionArgsList { args }),
                        _ => return Err(ParseError::ErrorParsingFunctionArgs),
                    }
                }

                Err(ParseError::ErrorParsingFunctionArgs)
            }
            _ => Err(ParseError::ErrorParsingFunctionArgs),
        }
    }

    fn func_arg(&mut self, name: &'a str) -> Result<FunctionArg<'a>, ParseError<'a>> {
        match self.peek_next_token()? {
            Some(Comma) => {
                self.step()?;
            }
            _ => {}
        }

        Ok(FunctionArg { name })
    }

    fn function_body(&mut self) -> Result<Vec<Statement<'a>>, ParseError<'a>> {
        let mut statements = Vec::new();

        match self.step()? {
            Some(IndentIncr) => {}
            _ => return Err(ParseError::IndentExpectedError),
        }

        while let Some(token) = self.peek_next_token()? {
            match token {
                IndentDecr => {
                    self.step()?;
                    return Ok(statements);
                }
                _ => statements.push(self.statement()?),
            }
        }

        Err(ParseError::UnexpectedEndOfInput)
    }

    fn expression(
        &mut self,
        right_binding_power: u32,
        first_token: Option<Token<'a>>,
    ) -> Result<Expression<'a>, ParseError<'a>> {
        let mut t = match first_token {
            Some(t) => t,
            None => self
                .step()?
                .map_or(Err(ParseError::UnexpectedEndOfInput), Ok)?,
        };

        let mut token = self.peek_next_token()?;

        let mut left = null_denotation(t);

        while right_binding_power < token.map(left_binding_power).unwrap_or(0) {
            t = token.map_or(Err(ParseError::UnexpectedEndOfInput), Ok)?;

            self.step()?;

            left = self.left_denotation(t, left)?;

            token = self.peek_next_token()?;
        }

        Ok(left)
    }

    fn left_denotation(
        &mut self,
        token: Token<'a>,
        left: Expression<'a>,
    ) -> Result<Expression<'a>, ParseError<'a>> {
        match token {
            BinOp(operator) => {
                let right = self.expression(left_binding_power(token), None)?;

                Ok(Expression::BinaryOp {
                    operator,
                    left: Box::new(left),
                    right: Box::new(right),
                })
            }
            _ => Ok(left),
        }
    }
}

fn null_denotation<'a>(token: Token<'a>) -> Expression<'a> {
    match token {
        Constant(c) => Expression::Constant(c),
        Name(name) => Expression::Variable(name),
        _ => panic!("unexpected token: {:?}", token),
    }
}

fn left_binding_power<'a>(token: Token<'a>) -> u32 {
    match token {
        BinOp(op) => op.binding_power(),
        _ => 0,
    }
}

#[derive(Debug, Copy, Clone)]
pub enum ParseError<'a> {
    TokeniserError(TokeniserError),
    UnexpectedToken(Token<'a>),
    UnexpectedEndOfInput,
    FunctionParseError,
    ErrorParsingFunctionArgs,
    IndentExpectedError,
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use std::fs;
    use test_case::test_case;

    #[test_case("src/fixtures/strings.lang"; "strings")]
    #[test_case("src/fixtures/maths.lang"; "maths")]
    #[test_case("src/fixtures/functions.lang"; "functions")]
    fn fixtures(fixture_file_name: &str) -> std::io::Result<()> {
        let contents = fs::read_to_string(fixture_file_name)?;

        let ast = parse(&contents);

        assert_debug_snapshot!(ast);

        Ok(())
    }
}
