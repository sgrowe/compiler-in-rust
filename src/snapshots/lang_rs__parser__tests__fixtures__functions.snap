---
source: src/parser.rs
expression: ast
---
Ok(
    Ast {
        statements: [
            FunctionDecl {
                name: "my_first_func",
                arguments: FunctionArgsList {
                    args: [],
                },
                body: [
                    Assignment {
                        name: "x",
                        expr: Constant(
                            Int(
                                5,
                            ),
                        ),
                    },
                    Assignment {
                        name: "y",
                        expr: BinaryOp {
                            operator: Multiply,
                            left: Variable(
                                "x",
                            ),
                            right: Constant(
                                Int(
                                    2,
                                ),
                            ),
                        },
                    },
                    BareExpression(
                        Variable(
                            "y",
                        ),
                    ),
                ],
            },
            FunctionDecl {
                name: "func_with_closure",
                arguments: FunctionArgsList {
                    args: [],
                },
                body: [
                    Assignment {
                        name: "x",
                        expr: Constant(
                            Int(
                                5,
                            ),
                        ),
                    },
                    FunctionDecl {
                        name: "top_stuff",
                        arguments: FunctionArgsList {
                            args: [],
                        },
                        body: [
                            BareExpression(
                                BinaryOp {
                                    operator: Multiply,
                                    left: Variable(
                                        "x",
                                    ),
                                    right: Constant(
                                        Int(
                                            2,
                                        ),
                                    ),
                                },
                            ),
                        ],
                    },
                    BareExpression(
                        Variable(
                            "top_stuff",
                        ),
                    ),
                ],
            },
            FunctionDecl {
                name: "function_with_arguments",
                arguments: FunctionArgsList {
                    args: [
                        FunctionArg {
                            name: "x",
                        },
                        FunctionArg {
                            name: "y",
                        },
                    ],
                },
                body: [
                    BareExpression(
                        BinaryOp {
                            operator: Multiply,
                            left: Variable(
                                "x",
                            ),
                            right: Variable(
                                "y",
                            ),
                        },
                    ),
                ],
            },
            FunctionDecl {
                name: "single_argument_func",
                arguments: FunctionArgsList {
                    args: [
                        FunctionArg {
                            name: "x",
                        },
                    ],
                },
                body: [
                    BareExpression(
                        BinaryOp {
                            operator: Plus,
                            left: BinaryOp {
                                operator: Multiply,
                                left: Constant(
                                    Int(
                                        2,
                                    ),
                                ),
                                right: Variable(
                                    "x",
                                ),
                            },
                            right: Constant(
                                Int(
                                    1,
                                ),
                            ),
                        },
                    ),
                ],
            },
        ],
    },
)
